File: update_scripts.py
#!/usr/bin/env python3
"""
Автоматическое обновление Apps Script из GitHub
Требует: pip install google-api-python-client google-auth-httplib2 google-auth-oauthlib requests
"""

import os
import json
import pickle
from pathlib import Path
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError
import requests
from dotenv import load_dotenv

# Загрузка переменных окружения
load_dotenv()

# Конфигурация
CONFIG = {
    'CLIENT_ID': os.getenv('GOOGLE_CLIENT_ID'),
    'CLIENT_SECRET': os.getenv('GOOGLE_CLIENT_SECRET'),
    'SCRIPT_ID': os.getenv('GOOGLE_SCRIPT_ID'),
    'GITHUB_REPO': os.getenv('GITHUB_REPO_URL'),
    'GITHUB_BRANCH': os.getenv('GITHUB_BRANCH', 'main'),
    'SCOPES': [
        'https://www.googleapis.com/auth/script.projects',
        'https://www.googleapis.com/auth/drive.file'
    ],
    'TOKEN_FILE': 'token.pickle',
    'CREDENTIALS_FILE': 'credentials.json'
}

# Файлы для обновления
FILES_TO_UPDATE = [
    'Config.gs',
    'Utils.gs',
    'Main.gs',
    'CurrencyManager.gs',
    'VATCalculator.gs',
    'PaymentManager.gs',
    'DebtCalculator.gs',
    'Notifications.gs',
    'Triggers.gs',
    'VersionManager.gs',
    'UpdateManager.gs',
    'MigrationScripts.gs',
    'CSVImporter.gs',
    'OptimizedSetup.gs'
]


def get_credentials():
    """Получить авторизованные credentials"""
    creds = None
    
    # Проверяем сохраненный токен
    if os.path.exists(CONFIG['TOKEN_FILE']):
        with open(CONFIG['TOKEN_FILE'], 'rb') as token:
            creds = pickle.load(token)
    
    # Если нет валидных credentials, запрашиваем авторизацию
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            if not os.path.exists(CONFIG['CREDENTIALS_FILE']):
                raise FileNotFoundError(
                    f"Credentials file not found: {CONFIG['CREDENTIALS_FILE']}\n"
                    "Please download OAuth credentials from Google Cloud Console"
                )
            
            flow = InstalledAppFlow.from_client_secrets_file(
                CONFIG['CREDENTIALS_FILE'],
                CONFIG['SCOPES']
            )
            creds = flow.run_local_server(port=0)
        
        # Сохраняем credentials для следующего раза с ограниченными правами
        with open(CONFIG['TOKEN_FILE'], 'wb') as token:
            pickle.dump(creds, token)
        
        # Устанавливаем права доступа: только владелец может читать/писать (Unix)
        import stat
        if os.name != 'nt':  # Не Windows
            os.chmod(CONFIG['TOKEN_FILE'], stat.S_IRUSR | stat.S_IWUSR)
    
    return creds


def fetch_from_github(file_name):
    """Загрузить файл из GitHub"""
    url = f"{CONFIG['GITHUB_REPO']}/raw/{CONFIG['GITHUB_BRANCH']}/{file_name}"
    
    response = requests.get(url)
    response.raise_for_status()
    
    return response.text


def update_script_file(service, file_name, content):
    """Обновить файл в Apps Script проекте"""
    script_name = file_name.replace('.gs', '')
    
    try:
        # Получаем содержимое проекта
        project = service.projects().getContent(scriptId=CONFIG['SCRIPT_ID']).execute()
        
        files = project.get('files', [])
        file_found = False
        
        # Ищем файл в проекте
        for file in files:
            if file.get('name') == script_name and file.get('type') == 'SERVER_JS':
                file_found = True
                file['source'] = content
                break
        
        if not file_found:
            # Создаем новый файл
            files.append({
                'name': script_name,
                'type': 'SERVER_JS',
                'source': content
            })
        
        # Обновляем проект
        service.projects().updateContent(
            scriptId=CONFIG['SCRIPT_ID'],
            body={'files': files}
        ).execute()
        
        print(f"✅ Updated: {file_name}")
        return True
        
    except HttpError as error:
        print(f"❌ Failed to update {file_name}: {error}")
        raise


def update_all_scripts():
    """Главная функция обновления"""
    try:
        print("🚀 Starting script update from GitHub...\n")
        
        # Проверка конфигурации
        if not all([CONFIG['CLIENT_ID'], CONFIG['CLIENT_SECRET'], CONFIG['SCRIPT_ID']]):
            raise ValueError("Missing configuration. Please check .env file")
        
        # Проверка безопасности - убеждаемся что не используем placeholder значения
        if ('your_client_id' in CONFIG['CLIENT_ID'] or 
            'your_client_secret' in CONFIG['CLIENT_SECRET'] or
            'your_script_id' in CONFIG['SCRIPT_ID']):
            raise ValueError("⚠️  SECURITY WARNING: Using placeholder values! Please set real credentials in .env file")
        
        # Проверка что credentials.json существует
        if not os.path.exists(CONFIG['CREDENTIALS_FILE']):
            raise FileNotFoundError(
                f"Credentials file not found: {CONFIG['CREDENTIALS_FILE']}\n"
                "Please download OAuth credentials from Google Cloud Console"
            )
        
        # Авторизация
        print("🔐 Authorizing...")
        creds = get_credentials()
        service = build('script', 'v1', credentials=creds)
        print("✅ Authorized\n")
        
        # Обновление файлов
        updated = 0
        failed = 0
        
        for file_name in FILES_TO_UPDATE:
            try:
                print(f"📥 Fetching {file_name}...")
                content = fetch_from_github(file_name)
                
                print(f"📝 Updating {file_name}...")
                update_script_file(service, file_name, content)
                
                updated += 1
            except Exception as error:
                print(f"❌ Error updating {file_name}: {error}")
                failed += 1
        
        print(f"\n✨ Update complete!")
        print(f"✅ Updated: {updated}")
        print(f"❌ Failed: {failed}")
        print(f"\n🔒 Security reminder: Never commit .env, credentials.json, or token.pickle to Git!")
        
    except Exception as error:
        print(f"💥 Fatal error: {error}")
        raise


if __name__ == '__main__':
    update_all_scripts()

File: update-scripts.js
/**
 * Автоматическое обновление Apps Script из GitHub
 * Требует: npm install googleapis google-auth-library
 */

const { google } = require('googleapis');
const { OAuth2Client } = require('google-auth-library');
const fs = require('fs').promises;
const path = require('path');
const https = require('https');

// Загрузка конфигурации из .env
require('dotenv').config();

const CONFIG = {
  CLIENT_ID: process.env.GOOGLE_CLIENT_ID,
  CLIENT_SECRET: process.env.GOOGLE_CLIENT_SECRET,
  SCRIPT_ID: process.env.GOOGLE_SCRIPT_ID,
  GITHUB_REPO: process.env.GITHUB_REPO_URL,
  GITHUB_BRANCH: process.env.GITHUB_BRANCH || 'main',
  TOKEN_PATH: path.join(__dirname, 'token.json'),
  CREDENTIALS_PATH: path.join(__dirname, 'credentials.json')
};

// Файлы для обновления
const FILES_TO_UPDATE = [
  'Config.gs',
  'Utils.gs',
  'Main.gs',
  'CurrencyManager.gs',
  'VATCalculator.gs',
  'PaymentManager.gs',
  'DebtCalculator.gs',
  'Notifications.gs',
  'Triggers.gs',
  'VersionManager.gs',
  'UpdateManager.gs',
  'MigrationScripts.gs',
  'CSVImporter.gs',
  'OptimizedSetup.gs'
];

/**
 * Загрузить файл из локальной папки
 */
async function fetchFromLocal(fileName) {
  const filePath = path.join(__dirname, fileName);
  try {
    const content = await fs.readFile(filePath, 'utf8');
    return content;
  } catch (error) {
    throw new Error(`File not found: ${filePath}`);
  }
}

/**
 * Загрузить файл из GitHub
 */
async function fetchFromGitHub(fileName) {
  if (!CONFIG.GITHUB_REPO || CONFIG.GITHUB_REPO.includes('your-username')) {
    throw new Error('GitHub repository not configured');
  }
  
  return new Promise((resolve, reject) => {
    const url = `${CONFIG.GITHUB_REPO}/raw/${CONFIG.GITHUB_BRANCH}/${fileName}`;
    
    https.get(url, (res) => {
      if (res.statusCode !== 200) {
        reject(new Error(`HTTP ${res.statusCode}: ${res.statusMessage}`));
        return;
      }
      
      let data = '';
      res.on('data', (chunk) => {
        data += chunk;
      });
      
      res.on('end', () => {
        resolve(data);
      });
    }).on('error', (err) => {
      reject(err);
    });
  });
}

/**
 * Загрузить файл (сначала из локальной папки, потом из GitHub)
 */
async function fetchFile(fileName) {
  // Сначала пробуем локальную папку
  try {
    return await fetchFromLocal(fileName);
  } catch (localError) {
    // Если локально не найдено, пробуем GitHub
    try {
      return await fetchFromGitHub(fileName);
    } catch (githubError) {
      throw new Error(`Failed to fetch ${fileName} from local or GitHub: ${localError.message}`);
    }
  }
}

/**
 * Авторизация в Google через OAuth2
 */
async function authorize() {
  const oAuth2Client = new OAuth2Client(
    CONFIG.CLIENT_ID,
    CONFIG.CLIENT_SECRET,
    'http://localhost:3000/oauth2callback'
  );
  
  // Проверяем сохраненный токен
  try {
    const token = await fs.readFile(CONFIG.TOKEN_PATH);
    const credentials = JSON.parse(token);
    oAuth2Client.setCredentials(credentials);
    
    // Проверяем, не истек ли токен
    if (credentials.expiry_date && credentials.expiry_date > Date.now()) {
      return oAuth2Client;
    }
    
    // Обновляем токен если истек
    const newToken = await oAuth2Client.refreshAccessToken();
    oAuth2Client.setCredentials(newToken.credentials);
    await fs.writeFile(CONFIG.TOKEN_PATH, JSON.stringify(newToken.credentials, null, 2));
    return oAuth2Client;
    
  } catch (err) {
    // Нет сохраненного токена - запрашиваем новый
    return getNewToken(oAuth2Client);
  }
}

/**
 * Получить новый токен
 */
async function getNewToken(oAuth2Client) {
  const authUrl = oAuth2Client.generateAuthUrl({
    access_type: 'offline',
    scope: [
      'https://www.googleapis.com/auth/script.projects',
      'https://www.googleapis.com/auth/drive.file'
    ],
  });
  
    console.log('\n🔐 Authorization required!');
    console.log('Visit this URL to authorize:', authUrl);
    console.log('Enter the code from that page here: ');
    console.log('⚠️  Keep this code private and do not share it!\n');
  
  // Для автоматизации можно использовать readline или встроенный сервер
  const readline = require('readline');
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  
  return new Promise((resolve) => {
    rl.question('Enter the code: ', async (code) => {
      rl.close();
      const { tokens } = await oAuth2Client.getToken(code);
      oAuth2Client.setCredentials(tokens);
      
    // Сохраняем токен с ограниченными правами доступа (только для владельца)
    await fs.writeFile(CONFIG.TOKEN_PATH, JSON.stringify(tokens, null, 2));
    
    // Устанавливаем права доступа: только владелец может читать/писать
    if (process.platform !== 'win32') {
      await fs.chmod(CONFIG.TOKEN_PATH, 0o600);
    }
    
    console.log('Token stored to', CONFIG.TOKEN_PATH);
      
      resolve(oAuth2Client);
    });
  });
}

/**
 * Обновить файл в Apps Script проекте
 */
async function updateScriptFile(auth, fileName, content) {
  const script = google.script({ version: 'v1', auth: auth });
  
  // Получаем имя файла без расширения
  const scriptName = fileName.replace('.gs', '');
  
  try {
    // Получаем список файлов проекта
    const project = await script.projects.getContent({
      scriptId: CONFIG.SCRIPT_ID
    });
    
    // Ищем файл в проекте
    let fileFound = false;
    const files = project.data.files || [];
    
    for (const file of files) {
      if (file.name === scriptName && file.type === 'SERVER_JS') {
        fileFound = true;
        // Обновляем содержимое файла
        file.source = content;
        break;
      }
    }
    
    if (!fileFound) {
      // Создаем новый файл
      files.push({
        name: scriptName,
        type: 'SERVER_JS',
        source: content
      });
    }
    
    // Обновляем проект
    await script.projects.updateContent({
      scriptId: CONFIG.SCRIPT_ID,
      requestBody: {
        files: files
      }
    });
    
    console.log(`✅ Updated: ${fileName}`);
    return true;
    
  } catch (error) {
    console.error(`❌ Failed to update ${fileName}:`, error.message);
    throw error;
  }
}

/**
 * Главная функция обновления
 */
async function updateAllScripts() {
  try {
    console.log('🚀 Starting script update...\n');
    
    // Проверка конфигурации
    if (!CONFIG.CLIENT_ID || !CONFIG.CLIENT_SECRET || !CONFIG.SCRIPT_ID) {
      throw new Error('Missing configuration. Please check .env file');
    }
    
    // Проверка безопасности - убеждаемся что не используем placeholder значения
    if (CONFIG.CLIENT_ID.includes('your_client_id') || 
        CONFIG.CLIENT_SECRET.includes('your_client_secret') ||
        CONFIG.SCRIPT_ID.includes('your_script_id')) {
      throw new Error('⚠️  SECURITY WARNING: Using placeholder values! Please set real credentials in .env file');
    }
    
    // Проверка что credentials.json существует
    try {
      await fs.access(CONFIG.CREDENTIALS_PATH);
    } catch (error) {
      throw new Error(`Credentials file not found: ${CONFIG.CREDENTIALS_PATH}\nPlease download OAuth credentials from Google Cloud Console`);
    }
    
    // Авторизация
    console.log('🔐 Authorizing...');
    const auth = await authorize();
    console.log('✅ Authorized\n');
    
    // Обновление файлов
    let updated = 0;
    let failed = 0;
    
    for (const fileName of FILES_TO_UPDATE) {
      try {
        console.log(`📥 Fetching ${fileName}...`);
        const content = await fetchFile(fileName);
        
        console.log(`📝 Updating ${fileName}...`);
        await updateScriptFile(auth, fileName, content);
        
        updated++;
      } catch (error) {
        console.error(`❌ Error updating ${fileName}:`, error.message);
        failed++;
      }
    }
    
    console.log(`\n✨ Update complete!`);
    console.log(`✅ Updated: ${updated}`);
    console.log(`❌ Failed: ${failed}`);
    
    // Предупреждение о безопасности
    console.log(`\n🔒 Security reminder: Never commit .env, credentials.json, or token.json to Git!`);
    
  } catch (error) {
    console.error('💥 Fatal error:', error.message);
    process.exit(1);
  }
}

// Запуск
if (require.main === module) {
  updateAllScripts();
}

module.exports = { updateAllScripts };

